loky/backend/utils.py

```diff
@@ -84,3 +94,207 @@ def _recursive_terminate(pid):
+            raise NotImplementedError("Could not find OpenBLAS library.")
+        self.lib_openblas.openblas_set_num_threads(num_threads)
+
+    def openblas_get_max_threads(self):
+        if self.lib_openblas is None:
+            raise NotImplementedError("Could not find OpenBLAS library.")
+        return self.lib_openblas.openblas_get_num_threads()
+
+    def _load_omp(self):
+        self.lib_gomp = self._load_lib(OMP_MODULE_NAME_GNU, "")
+        self.lib_iomp = self._load_lib(OMP_MODULE_NAME_INTEL, "")
+
+    def openmp_gnu_set_num_threads(self, num_threads):
+
+        if self.lib_gomp is None:
+            raise NotImplementedError("Could not find OpenMP library")
+        self.lib_gomp.omp_set_num_threads(num_threads)
+
+    def openmp_gnu_get_max_threads(self):
+        if self.lib_gomp is None:
+            raise NotImplementedError("Could not find OpenMP library")
+        return self.lib_gomp.omp_get_max_threads()
+
+    def openmp_intel_set_num_threads(self, num_threads):
+        if self.lib_iomp is None:
+            raise NotImplementedError("Could not find OpenMP library")
+        self.lib_iomp.omp_set_num_threads(num_threads)
+
+    def openmp_intel_get_max_threads(self):
+        if self.lib_iomp is None:
+            raise NotImplementedError("Could not find OpenMP library")
+        return self.lib_iomp.omp_get_max_threads()
+
+    def _load_mkl(self):
+        self.lib_mkl = self._load_lib(MKL_MODULE_NAME, "blas_mkl_info")
+
+    def mkl_set_num_threads(self, num_threads):
+        if self.lib_mkl is None:
+            raise NotImplementedError("Could not find MKL libray")
+        self.lib_mkl.MKL_Set_Num_Threads(num_threads)
+
+    def mkl_get_max_threads(self):
+        if self.lib_mkl is None:
+            raise NotImplementedError("Could not find MKL libray")
+        return self.lib_mkl.MKL_Get_Max_Threads()
+
+    def _load_lib(self, module_name, lib_info):
+        lib_name = find_library(module_name)
+        if lib_name is not None:
+            return ctypes.CDLL(lib_name, use_errno=True)
+        return self._get_lib_from_dynamic_libs(module_name)
+
+    def _get_lib_from_dynamic_libs(self, module_name):
+
+        global _MODULE_NAME
+        _MODULE_NAME = None
+
+        libc_name = find_library("c")
+        if libc_name is None:
+            return
+        libc = ctypes.CDLL(libc_name)
+        if not hasattr(libc, "dl_iterate_phdr"):
+            return
+
+        signature_cfunc = ctypes.CFUNCTYPE(
+            ctypes.c_int, ctypes.c_void_p, ctypes.c_size_t, ctypes.c_char_p)
+        c_check_module = signature_cfunc(check_module)
+
+        data = ctypes.c_char_p(module_name.encode('utf-8'))
+        res = libc.dl_iterate_phdr(c_check_module, data)
+        if res == 1:
+            return ctypes.CDLL(_MODULE_NAME)
+
+
+_clibs_wrapper = None
+
+
+def _get_wrapper():
+    global _clibs_wrapper
+    if _clibs_wrapper is None:
+        _clibs_wrapper = _CLibsWrapper()
+    return _clibs_wrapper
```

Comment:
This singleton / factory mechanism that only perform the library lookup in the init of the singleton whill make it impossible to introspect and reconfigure the threading runtimes if they are initialized in a delayed manner, for instance when importing numpy as is the case for OpenBLAS:

```
$ python -c "from loky.backend.utils import get_thread_limits; get_thread_limits(); import numpy; print(get_thread_limits()['openblas'])"
None
$ python -c "from loky.backend.utils import get_thread_limits; import numpy; print(get_thread_limits()['openblas'])"
12
```

Reply:
We could try to load library that are `None` each time we call a function from the class to avoid this but it means more overhead.